### ISSUE TITLE
Form Constants or Hidden Fields.

### ISSUE THREAD

**Issue by kyse** [json-schema-form/angular-schema-form#674]
I have a schema that includes a few array properties that I need to include a property on the array's contents objects.  The property differs depending on which object in the array it is.

For example:

Schema:
```
{
  type: 'object',
  title: 'MyObj',
  properties: {
    PointOfContact: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          isPrimary: {
            type: 'boolean',
          },
          Name: {
            type: 'string'
          }
        }
      }
    }
  }
}
```

So for my form, I need to display primary POC and alt POC fields.

Form:
```
[
  {
    key: 'PointOfContact[0].Name'
  },
  {
    key: 'PointOfContact[0].isPrimary'
  },
  {
    key: 'PointOfContact[1].Name'
  },
  {
    key: 'PointOfContact[1].isPrimary'
  }
]
```

I need my model's PointOfContact[0].isPrimary to = true, and false otherwise.  I'd rather not have this field displayed because it doesn't make sense to do so.  I have similar cases where objects in the array need to be displayed as fields in the form, but for the model, I need to include a specific foreign key value for that field's target object.

My understanding is, the point of schema forms is to reduce the need for doing things manually, so the lack of an ability to apply a value to the modal without rendering it on the form, or supplying a hidden input type even is boggling.

I guess I could create a new form type 'hidden' to supply the additional values I need to my model?

So my question to the community:  Am I approaching this wrong?  What am I missing here?

**Comment by Anthropic** [comment]
Where does is primary get set? Is there a reason you can't set the ID for which item is primary outside the object and reference it within the array items?

**Comment by kyse** [comment]
So, for that example, isPrimary would get set in my form definition since I'm manually defining the fields for that array object at index 0 to be the primary POC.  Since this scenario would be useful for more than just a on or off bit, but specifying foreign keys like a typeid of a lookupvalue table, a hidden field where I can pass in the value would be nice to have.

This way I could setup an engine to parse template's based on some kind of roles/permissions to include in the form and build out the form template based on the schema/odata metadata automatically.

Yes, I could just go into the model and apply the properties manually from the controller's onSubmit method, but that seems counter to the whole point of schema forms.  Also, splits the schema/form logic into a third place.

**Comment by Anthropic** [comment]
@kyse I think I see your issue, there is a PR related to acting upon array using arrayIndeces, but in extension to that there should be an easier way to set a field default in the form that is calculated. Even using a calculation combined with a destroyStrategy doesn't appear to work as I expected for your purposes. I'll look into it.

**Comment by kyse** [comment]
Yea, this is where I am at so far.

Schema:
```
                {
                    type: 'object',
                    title: 'Requestor',
                    properties: {
                        People: {
                            type: 'array',
                            default: [],
                            items: {
                                type: 'object',
                                properties: {
                                    FirstName: {
                                        title: 'First Name',
                                        type: 'string',
                                    },
                                    LastName: {
                                        title: 'Last Name',
                                        type: 'string',
                                    },
                                    Title: {
                                        title: 'Title',
                                        type: 'string',
                                    },
                                    PhoneNumber: {
                                        title: 'Phone Number',
                                        type: 'string',
                                    },
                                    Email: {
                                        title: 'E-Mail Address',
                                        type: 'string',
                                        pattern: '^.+@.+[.].+$',
                                    },
                                    isPrimary: {
                                        type: 'boolean',
                                    },
                                }
                            }
                        },
                        AdditionalInfoes: {
                            type: 'array',
                            default: [],
                            items: {
                                type: 'object',
                                properties: {
                                    Information: {
                                        title: 'Info',
                                        type: 'string',
                                    },
                                    TypeID: {
                                        type: 'number',
                                    }
                                },
                            }
                        },
                    },
                };
```

Form Def (I'm using extension methods on an array to build out the array, method(key, class, additionalParamsToMerge).  You'll notice a pattern that can be turned into a build engine.):
```
                // Create the tabs
                var tabRequestor = tabs.tab('Requestor');

                var pocPrimary = tabRequestor.fieldset('Primary POC', 'col-xs-6');
                pocPrimary.hidden('People[0].isPrimary', true); // key, value
                pocPrimary.text('People[0].FirstName', null, { title: 'First Name'}); // key, htmlClass, otherOptionsObject
                pocPrimary.text('People[0].LastName', null, { title: 'Last Name' });
                pocPrimary.text('People[0].Title', null, { title: 'Title' });
                pocPrimary.text('People[0].Email', null, { title: 'E-Mail Address' });
                pocPrimary.text('AdditionalInfoes[0].Information', null, { title: 'Other Email 1' });
                pocPrimary.hidden('AdditionalInfoes[0].TypeID', 13, null, { condition: '!!model.AdditionalInfoes[0].Information' });
                pocPrimary.text('AdditionalInfoes[1].Information', null, { title: 'Other Email 2' });
                pocPrimary.hidden('AdditionalInfoes[1].TypeID', 15, null, { condition: '!!model.AdditionalInfoes[1].Information' });
                pocPrimary.text('People[0].PhoneNumber', null, { title: 'Phone Number' });
                pocPrimary.text('AdditionalInfoes[2].Information', null, { title: 'Other Phone Number 1' });
                pocPrimary.hidden('AdditionalInfoes[2].TypeID', 3, null, { condition: '!!model.AdditionalInfoes[2].Information' });
                pocPrimary.text('AdditionalInfoes[3].Information', null, { title: 'Other Phone Number 2' });
                pocPrimary.hidden('AdditionalInfoes[3].TypeID', 5, null, { condition: '!!model.AdditionalInfoes[3].Information' });
                pocPrimary.text('AdditionalInfoes[4].Information', null, { title: 'Fax Number' });
                pocPrimary.hidden('AdditionalInfoes[4].TypeID', 7, null, { condition: '!!model.AdditionalInfoes[4].Information' });
                pocPrimary.text('AdditionalInfoes[5].Information', null, { title: 'Other Fax Number 1' });
                pocPrimary.hidden('AdditionalInfoes[5].TypeID', 9, null, { condition: '!!model.AdditionalInfoes[5].Information' });
                pocPrimary.text('AdditionalInfoes[6].Information', null, { title: 'Other Fax Number 2' });
                pocPrimary.hidden('AdditionalInfoes[6].TypeID', 11, null, { condition: '!!model.AdditionalInfoes[6].Information' });

                var pocAlt = tabRequestor.fieldset('Alternate POC', 'col-xs-6');
                pocAlt.hidden('People[1].isPrimary', false);
                pocAlt.text('People[1].FirstName', null, { title: 'First Name' });
                pocAlt.text('People[1].LastName', null, { title: 'Last Name' });
                pocAlt.text('People[1].Title', null, { title: 'Title' });
                pocAlt.text('People[1].Email', null, { title: 'E-Mail Address' });
                pocAlt.text('AdditionalInfoes[7].Information', null, { title: 'Other Email 1' });
                pocAlt.hidden('AdditionalInfoes[7].TypeID', 14, null, { condition: '!!model.AdditionalInfoes[7].Information' });
                pocAlt.text('AdditionalInfoes[8].Information', null, { title: 'Other Email 2' });
                pocAlt.hidden('AdditionalInfoes[8].TypeID', 16, null, { condition: '!!model.AdditionalInfoes[8].Information' });
                pocAlt.text('People[1].PhoneNumber', null, { title: 'Phone Number' });
                pocAlt.text('AdditionalInfoes[9].Information', null, { title: 'Other Phone Number 1 });
                pocAlt.hidden('AdditionalInfoes[9].TypeID', 4, null, { condition: '!!model.AdditionalInfoes[9].Information' });
                pocAlt.text('AdditionalInfoes[10].Information', null, { title: 'Other Phone Number 2' });
                pocAlt.hidden('AdditionalInfoes[10].TypeID', 6, null, { condition: '!!model.AdditionalInfoes[10].Information' });
                pocAlt.text('AdditionalInfoes[11].Information', null, { title: 'Fax Number' });
                pocAlt.hidden('AdditionalInfoes[11].TypeID', 8, null, { condition: '!!model.AdditionalInfoes[11].Information' });
                pocAlt.text('AdditionalInfoes[12].Information', null, { title: 'Other Fax Number 1' });
                pocAlt.hidden('AdditionalInfoes[12].TypeID', 10, null, { condition: '!!model.AdditionalInfoes[12].Information' });
                pocAlt.text('AdditionalInfoes[13].Information', null, { title: 'Other Fax Number 2' });
                pocAlt.hidden('AdditionalInfoes[13].TypeID', 12, null, { condition: '!!model.AdditionalInfoes[13].Information' });
```

Template "hidden.html":
```
    $templateCache.put("decorators/bootstrap/hidden.html", "<div class=\"form-group schema-form-{{form.type}} {{form.htmlClass}}\"><input type=\"{{form.type}}\" step=\"any\" sf-changed=\"form\" class=\"form-control {{form.fieldHtmlClass}}\" id=\"{{form.key.slice(-1)[0]}}\" schema-validate=\"form\" sf-field-model=\"replaceAll\" ng-model=\"$$value$$\" name=\"{{form.key.slice(-1)[0]}}\" aria-describedby=\"{{form.key.slice(-1)[0] + \'Status\'}}\" value=\"form.value\" ng-init=\"$$value$$=form.value\"></div>");

   function configSchemaForm(schemaFormProvider, schemaFormDecoratorsProvider, sfBuilderProvider, sfPathProvider) {
        var base = 'decorators/bootstrap/';

        var simpleTransclusion = sfBuilderProvider.builders.simpleTransclusion;
        var ngModelOptions = sfBuilderProvider.builders.ngModelOptions;
        var ngModel = sfBuilderProvider.builders.ngModel;
        var sfField = sfBuilderProvider.builders.sfField;
        var condition = sfBuilderProvider.builders.condition;
        var array = sfBuilderProvider.builders.array;

        var defaults = [sfField, ngModel, ngModelOptions, condition];

        schemaFormDecoratorsProvider.defineAddOn('bootstrapDecorator', 'hidden', base + 'hidden.html', defaults);
    }
```

The end result of this in my model object is:
People property is an object instead of an array (why?) containing 0 and 1, with isPriority set to true and false properly.  Other fields applied properly.

AdditionalInfoes property is an array as expected.  TypeID is applied only to proper object in array when and only when Information field is populated.  Array length is size of last element populated (ie if I only fill in field Other Phone Number 1 for primary POC I get an array size 3, but only array[2] is in the array, indexes 0 and 1 are missing.  Also, when typing into a text type field then clearing it out, it remains on my model as an empty string.  I guess there are use cases for this?  Would be nice to add an form level default option for this to remove the property and in this case since its contained in an array object, the object itself from the array if it its empty after removing the field property, rather than setting the property to "".  I looked at destroySolution for this but I think it's only related to the specific array form type builder and removing the object from the array via the remove button.  Not properties themselves.

Hopefully this wont be an issue on post, but I guess I could just do a Array.prototype.slice on it to reset the array indexes before post.

Then again, this really isn't a solution since the values wont bind to the proper fields if I set model = the object from the OData api response.  I need to expand this to conditionally apply additionalinfoe's information property title based on the typeid somehow for an existing model.  If I just use AdditionalInfoes[].Information, all those fields get the same result since the controller isn't considering it an array.  Perhaps I need to create a text template using the array builder?

I was looking at destroyStrategy originally for removing the typeID's before I started messing with conditions and noticed they seem to only apply when removing an entity from an array, but in this case, again, not using the array builder with this setup, just hacking the fields into an array model key, which sometimes just ends up being an object instead of an array.

After typing all this and thinking through my thoughts I'm wondering if its just better to leave the logic of building the final post object and parsing an existing entity into a form model schema forms can bind to more appropriate as far as time constraints goes.  It just seems, if the API is smart enough to take the post object and apply updates internally to the proper tables, that moves the onus of supporting such complex forms that span multiple tables to schema forms to support.  And by that I in no way mean to degrade schema forms and what you guys have created so far.  It's amazing work and I applaud it.  Just thinking improvement wise.  Maybe if you agree I'm going in the right direction to resolve my scenario I can spend some personal time creating a pull requests to resolve.  Just don't want to waste time on a direction thats not in your vision for your library.

Cheers!

**Comment by Anthropic** [comment]
@kyse could you make a plunker possibly? That's a lot to follow outside my work hours, a plunker would help my visualise what you're trying to do better.

**Comment by kyse** [comment]
Sure, does this link work for you? [https://plnkr.co/NsEzQGz5DoRdD018RTWn](https://plnkr.co/NsEzQGz5DoRdD018RTWn)

**Comment by kyse** [comment]
Just an additional note. If I use condition: 'model.People[0]' or 'model.People[1]' on the two hidden IsPrimary people fields I get People as an array in the model, but my desired result is to always have an array of 2 People entities in the array with that IsPrimary field populated.  So even though that could be a solution, it defeats the purpose of using the hidden input field as a way to apply constants in this scenario.

**Comment by kyse** [comment]
Updated the plunker to resolve the array being initialized as an object issue.  Using default containing the default values I want (non conditional) resolves that problem.  Just need to figure out how to bind the AdditionalInfo fields to the array without hard coding the index.  As I can't be sure the array being returned from the API will be in the proper order, since they aren't required to include all fields, the current setup wont work.  Maybe I need to write some custom binding logic for arrays into the text directive?  Need to be able to find a field's key in array format to the array where a contained entity contains a specific key, or keys to take care of binding from the model to the field.  And handle properly adding a new entity to the array without adding a bunch of nulls (an effect of me currently manually binding to an array index for each field).

Also, still trying to figure out how to manipulate ngModel to delete the entity rather than leaving an empty string when the field is cleared.  Actually deleted not cleared.  I noticed when hitting the x (in IE) in the checkbox to clear the checkbox it doesn't trigger the onChanged event, and the model retains the uncleared value.

**Comment by Anthropic** [comment]
@kyse there is work needed to fix how arrays are behaving, probably our biggest pain point is that the library does use ngModel, would probably have been better to avoid it given how many quirks it has. But too late now. I really want to help you more, please don't take my slow responses the wrong way! But I spent almost an hour looking at your problem before realising I really need to focus on updating the library and starting to fix some of the underlying issues that are responsible.

I can see your issues and once the lib core is updated we can start tackling them, but for now I would really suggest using a struct `poc:{primary:{},secondary:{}}` as the easiest way to overcome the issue. An onsubmit could loop the keys into the array structure you desire.

The delete issue can partially be overcome with destroy strategy but can only do what ngModel will let it, I think there are already issues raised about that.

**Comment by Anthropic** [comment]
Oh and nice plunker, helped a lot to visualise what you are trying to achieve, thank you, when I get around to fixing the underlying issues it will really help! :)

**Comment by kyse** [comment]
Thanks for your time on this.  I've just about come to the same conclusion.  Thinking about one last avenue before I give up.  What about being able to provide a getter/setter function to ngModel that allows control over assigning a field to a new entity in an array based on a condition.  So if setting, it would find the entity (or create an index lookup store) and update, or push a new entity to the target array.  If getting, it would use the index hash lookup table to retrieve the proper array index.

So to summarize:
1) Global form ngModelOptions, set getterSetter to true.
2) Append the model or array with the appropriate getter/setter functions needed for the fields in the array.
3) Come up with some form field definition option for passing conditional binding requirements inside the getter/setter function.

Might end up being too much trouble, and the hack in the submit method might be the easiest. :/

**Comment by kyse** [comment]
So this is what I've ended up with thus far to take control of applying fields to entities inside an array conditionally.

```
          arrayKey: function arrayKey(arrName, fieldName, condition) {
                if (!angular.isDefined(condition) || (!angular.isFunction(condition) && !angular.isString(condition)))
                    console.error('Invalid condition parameter passed to arrayKey.', condition);
                if (!this.vm.model.$$gs)
                    Object.defineProperty(this.vm.model, '$$gs', {
                        writable: true,
                        configurable: true,
                        enumerable: false,
                        value: {},
                    });
                var model = this.vm.model;
                var gsName = [arrName, fieldName].join('-');
                var conditionFn = condition ? angular.isFunction(condition) ? condition : conditionFilter : undefined;
                model.$$gs[gsName] = arrayHandler;

                function arrayHandler(value) {
                    // Double check here if array exists and create?
                    var found = model[arrName].filter(conditionFn),
                        entity;
                    if (found.length) {
                        entity = found[0];
                    } else {
                        entity = {};
                        model[arrName].push(entity);
                    }
                    return arguments.length ? (value === '' ? entity[fieldName] = undefined : entity[fieldName] = value) : entity[fieldName];
                }

                function conditionFilter(entity) {
                    var result = true;

                    // Assume string = comma seperated property names, and converted boolean value, prefix propertyName with ! to do negative eval.
                    if (angular.isString(condition)) {
                        angular.forEach(condition.split(','), function (property) {
                            var success;
                            switch (property[0]) {
                                case '!':
                                    success = !entity[property.substring(1)];
                                    break;
                                default:
                                    success = !!entity[property];
                                    break;
                            }
                            result = result && success;
                        }, this);
                        return result;
                    }

                    // Assume kvp object ( Property Name prefixed with ! is '!==', = is '===', else '=='
                    angular.forEach(condition, function (value, name) {
                        var success;
                        switch(name[0]) {
                            case '!':
                                success = entity[name.substring(1)] !== value;
                                break;
                            case '=':
                                success = entity[name.substring(1)] === value;
                                break;
                            default:
                                success = entity[name] == value;
                                break;
                        }
                        result = result && success;
                    }, this);

                    return result;
                }

                return '$$gs.' + gsName;
            }
```

Caveats:

- Create model first before defining the form definition array.
- When you define your forms, use this function to return the form def's key property value.
- So far I've tested it where the array entities already exist via schema's default: [{conditionProperty: conditionValue}].  Will probably have to take over handling the creation of the entity in the array if it doesn't exist, which means taking into consideration any default values needed.  Hopefully this should just work for hidden field keys as well.
- Finally, since it's targeting a key that doesn't have a schema defined I've noticed the field doesn't get ngModelOptions from the global form defaults set.  This problem also existed when I target an array manually (ie People[0].FirstName).  So you'll have to include in your field def, `ngModelOptions: { getterSetter: true }`.
- You can take over full control of how the fields get applied/removed from the model, ie instead of leaving an empty string on the model when a user deletes everything out of the field, you can set it to undefined (awesome!).

Would probably make sense to turn this into an optional builder to detect some syntax in the key targeting an array and apply the required conditional option string or function to a prebuilt method for retrieving the entity from the array.  Not sure if that can all be handled from the builder though since you really need access to the model to apply changes.  This would give better control/options for destroy strategy if applied everywhere, not sure how bad it'll be performance wise though on large forms.  Since I am building out a large form, we shall see.  It's only really built for the fields targeting arrays though.

**Comment by kyse** [comment]
This has evolved into creating a new directive for schema forms for me.  I've created a getterSetter builder that applies an sf-gs attribute to the ng-model targeted element.  The gs method handles initializing the array and updating the entity with the new values.  Needs to be expanded to handle removing the entities from the array if null, etc but this is a good POC to get the point accross.

[Plunker: Schema Forms - GetterSetter POC](https://plnkr.co/IA48cewtE24GAvMX3TlN)

I feel like this route could add configurable settings to better handle adding or removing all inputs from the model, ie remove if empty string, remove entity object from array if empty when property is removed, etc.  Even better if ngModel allows stacking getterSetter methods, could apply this logic as a standard for the ngModel builder, and apply the gs builder on top for specific array binding cases.

My only real concern is performance.  But so far it seems to be working good for me on a form containing over 100 fields with 40 of them being array bound via the gs builder.

Thoughts?

**Comment by krptodr** [comment]
None of these Plunkers exist anymore. Can anyone please provide a new one?

**Comment by jeff-emed** [comment]
please?

**Comment by kyse** [comment]
I logged in and made it public, not sure why those settings changed.  I dont have to time to confirm it still works, my guess probably an issue with vendor script import links to get it fixed up and in working order again, but at least you can see the implementation code.

https://plnkr.co/plunk/IA48cewtE24GAvMX3TlN
